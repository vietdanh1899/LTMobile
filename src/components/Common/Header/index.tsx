import React, { PureComponent } from 'react';
import {
  StyleSheet, Image,
} from 'react-native';
import {
  withTheme, ThemeProps,
  Header as EHeader,
  HeaderProps as EHeaderProps,
  Icon,
} from 'react-native-elements';
import NavigationService from '@utils/navigation';
import SwitchChangeTheme from '@contents/Config/Shared/SwitchChangeTheme';
import { headerHeight } from '@themes/ThemeComponent/Common/CommonProps';
import QuickView from '../View/QuickView';

export interface HeaderProps extends EHeaderProps {
  height?: number | string;
  width?: number | string;
  position?: 'relative' | 'absolute';
  top?: number;
  left?: number;
  right?: number;
  bottom?: number;
  backgroundColor?: string;
  borderBottomColor?: string;
  borderBottomWidth?: number;
  transparent?: boolean;
  backIcon?: boolean;
  leftIconBackgroundColor?: string;
  closeIcon?: boolean;
  title?: string;
  logo?: boolean;
  switchTheme?: boolean;
  shadow?: boolean;
  leftColor?: string;
  centerColor?: string;
  rightColor?: string;
  color?: string;
  theme?: any;
}

const styles = StyleSheet.create({
  defaultTitleStyle: {
    fontWeight: 'bold',
    fontSize: 22,
  },
  defaultLogoStyle: {
    width: 84,
    height: 30,
    resizeMode: 'contain',
  },
});

class Header extends PureComponent<HeaderProps> {
  static defaultProps = {
    placement: 'center',
    leftIconBackgroundColor: 'transparent',
    height: headerHeight,
    width: '100%',
  };

  render() {
    const {
      position,
      height,
      width,
      top,
      left,
      right,
      bottom,
      backgroundColor: backgroundColorProp,
      transparent,
      backIcon,
      leftIconBackgroundColor,
      closeIcon,
      title,
      leftColor: leftColorProp,
      centerColor: centerColorProp,
      rightColor: rightColorProp,
      color: colorProp,
      borderBottomColor,
      borderBottomWidth,
      placement,
      logo,
      containerStyle: containerStyleProp,
      leftContainerStyle: leftContainerStyleProp,
      leftComponent: leftComponentProp,
      centerContainerStyle: centerContainerStyleProp,
      centerComponent: centerComponentProp,
      rightContainerStyle: rightContainerStyleProp,
      rightComponent: rightComponentProp,
      shadow,
      switchTheme,
      theme,
      ...otherProps
    } = this.props;

    const backgroundColor = transparent ? 'transparent' : (backgroundColorProp || theme.Header.backgroundColor);
    let leftColor = leftColorProp || theme.Header.leftColor;
    let centerColor = centerColorProp || theme.Header.centerColor;
    let rightColor = rightColorProp || theme.Header.rightColor;
    if (colorProp) {
      leftColor = colorProp;
      centerColor = colorProp;
      rightColor = colorProp;
    }

    const containerStyle = StyleSheet.flatten([
      {
        position,
        height,
        width,
        top,
        left,
        right,
        bottom,
        borderBottomColor,
        borderBottomWidth,
        backgroundColor,
        paddingBottom: 10,
        paddingHorizontal: theme.headerPaddingHorizontal,
      },
      shadow ? [{ marginBottom: 3 }, theme.shadowView] : undefined,
      containerStyleProp,
    ]);

    /**
     * Left Component
     */
    const leftContainerStyle = StyleSheet.flatten([
      {
        paddingRight: ((backIcon || closeIcon)) ? 10 : 0,
      },
      leftContainerStyleProp,
    ]);
    let leftComponent: any = leftComponentProp;
    if (backIcon) {
      leftComponent = {
        icon: 'arrowleft',
        type: 'antdesign',
        size: 25,
        color: leftColor,
        onPress: () => NavigationService.goBack(),
        style: {
          width: 25, height: 25,
        },
        containerStyle: leftIconBackgroundColor !== 'transparent' ? {
          padding: 8, backgroundColor: leftIconBackgroundColor, borderRadius: 20,
        } : null,
      };
    } else if (closeIcon) {
      leftComponent = {
        icon: 'close',
        type: 'antdesign',
        size: 25,
        color: leftColor,
        onPress: () => NavigationService.goBack(),
        style: {
          width: 25, height: 25,
        },
        containerStyle: leftIconBackgroundColor !== 'transparent' ? {
          padding: 8, backgroundColor: leftIconBackgroundColor, borderRadius: 20,
        } : null,
      };
    }

    /**
     * Center Component
     */
    let centerComponent: any = centerComponentProp;
    if (title) {
      centerComponent = {
        text: title,
        style: StyleSheet.flatten([
          styles.defaultTitleStyle,
          {
            marginLeft: backIcon || closeIcon ? -10 : 0,
            color: centerColor,
          },
          centerContainerStyleProp,
        ]),
      };
    } else if (logo) {
      centerComponent = (
        <QuickView testID="centerComponentLogo" marginLeft={(backIcon && placement === 'center') || (closeIcon && placement === 'center') ? -20 : 0}>
          <Image
            style={styles.defaultLogoStyle}
            source={require('@images/headerLogo.png')}
          />
        </QuickView>
      );
    }

    /**
     * Right Component
     */
    let rightComponent: any = rightComponentProp;
    if (switchTheme) {
      rightComponent = (
        <QuickView testID="rightComponentSwitchTheme" row center>
          <Icon name="theme-light-dark" type="material-community" style={{ marginRight: 5 }} color={rightColor} />
          <SwitchChangeTheme />
        </QuickView>
      );
    }

    return (
      <EHeader
        {...otherProps}
        placement={placement}
        leftComponent={leftComponent}
        centerComponent={centerComponent}
        rightComponent={rightComponent}
        containerStyle={containerStyle}
        leftContainerStyle={leftContainerStyle}
        ViewComponent={QuickView}
        linearGradientProps={{ backgroundColor: 'blue' }}
      />
    );
  }
}

export default withTheme(Header as React.ComponentType<HeaderProps & ThemeProps<any>>);
