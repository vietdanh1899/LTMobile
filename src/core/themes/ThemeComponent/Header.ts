import { Color, lightPrimaryColor, darkPrimaryColor } from './Common/Color';

/**
 * Header
 */
export const darkHeader = {
  backgroundColor: darkPrimaryColor,
  leftColor: Color.black,
  centerColor: Color.black,
  rightColor: Color.black,
};
export const lightHeader = {
  backgroundColor: lightPrimaryColor,
  leftColor: Color.white,
  centerColor: Color.white,
  rightColor: Color.white,
};
