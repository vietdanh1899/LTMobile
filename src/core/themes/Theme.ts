import {
  bottomNavigationBarHeight,
  headerPaddingHorizontal,
  bodyPaddingHorizontal,
  shadowViewLight,
  shadowViewDark,
  darkComponentColor,
  lightComponentColor,
} from './ThemeComponent/Common/CommonProps';
import {
  Color,
  darkPlatform,
  lightPlatform,
  darkPrimaryColor,
  darkSecondaryColor,
  lightPrimaryColor,
  lightSecondaryColor,
} from './ThemeComponent/Common/Color';
import { darkText, lightText } from './ThemeComponent/Text';
import { darkButton, lightButton } from './ThemeComponent/Button';
import { darkHeader, lightHeader } from './ThemeComponent/Header';
import { darkInput, lightInput } from './ThemeComponent/Input';
import { darkCheckBox, lightCheckBox } from './ThemeComponent/ListCheckBox';
import { darkDropdown, lightDropdown } from './ThemeComponent/Dropdown';
import { darkModal, lightModal } from './ThemeComponent/Modal';
import { lightParallaxScrollView, darkParallaxScrollView } from './ThemeComponent/ParallaxScrollView';

const ThemeEnum = {
  DARK: 'dark',
  LIGHT: 'light',
};

/**
 * Theme
 */
export const darkTheme = {
  key: ThemeEnum.DARK,
  colors: {
    primary: darkPrimaryColor,
    secondary: darkSecondaryColor,

    white: Color.white,
    grey1: Color.grey1,
    grey2: Color.grey2,
    grey3: Color.grey3,
    grey4: Color.grey4,
    grey5: Color.grey5,
    grey6: Color.grey6,
    grey7: Color.grey7,
    grey8: Color.grey8,
    grey10: Color.grey10,
    black: Color.black,

    greyOutline: '#bbb',
    searchBg: '#303337',

    success: Color.green,
    error: Color.red,
    warning: Color.yellow,
    disabled: 'hsl(208, 8%, 90%)',
    // Darker color if hairlineWidth is not thin enough
    divider: Color.grey6,
    border: Color.grey9,
    platform: darkPlatform,

    /**
    * Component Color
    */
    bgColor: darkComponentColor.bgColor,
    bgColorSecondary: darkComponentColor.bgColorSecondary,
    textColor: darkComponentColor.textColor,
    textColorSecondary: darkComponentColor.textColorSecondary,
  },
  /**
   * Component
   */
  bottomNavigationBarHeight,
  headerPaddingHorizontal,
  bodyPaddingHorizontal,
  shadowView: shadowViewDark,

  Text: darkText,
  Button: darkButton,
  Header: darkHeader,
  Input: darkInput,
  CheckBox: darkCheckBox,
  Dropdown: darkDropdown,
  AppModal: darkModal,
  ParallaxScrollView: darkParallaxScrollView,
};

export const lightTheme = {
  key: ThemeEnum.LIGHT,
  colors: {
    primary: lightPrimaryColor,
    secondary: lightSecondaryColor,

    white: Color.white,
    grey1: Color.grey1,
    grey2: Color.grey2,
    grey3: Color.grey3,
    grey4: Color.grey4,
    grey5: Color.grey5,
    grey6: Color.grey6,
    grey7: Color.grey7,
    grey8: Color.grey8,
    grey10: Color.grey10,
    black: Color.black,

    greyOutline: '#bbb',
    searchBg: '#303337',

    listItemBg: Color.white,
    success: Color.green,
    error: Color.red,
    warning: Color.yellow,
    disabled: 'hsl(208, 8%, 90%)',
    // Darker color if hairlineWidth is not thin enough
    divider: Color.grey2,
    border: Color.grey2,
    platform: lightPlatform,

    /**
     * Component Color
     */
    bgColor: lightComponentColor.bgColor,
    bgColorSecondary: lightComponentColor.bgColorSecondary,
    textColor: lightComponentColor.textColor,
    textColorSecondary: lightComponentColor.textColorSecondary,
  },

  /**
   * Component
   */
  bottomNavigationBarHeight,
  headerPaddingHorizontal,
  bodyPaddingHorizontal,
  shadowView: shadowViewLight,

  Text: lightText,
  Button: lightButton,
  Header: lightHeader,
  Input: lightInput,
  CheckBox: lightCheckBox,
  Dropdown: lightDropdown,
  AppModal: lightModal,
  ParallaxScrollView: lightParallaxScrollView,
};

export { Color };
export default lightTheme;
