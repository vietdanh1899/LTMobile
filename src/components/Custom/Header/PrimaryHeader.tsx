import React, { PureComponent } from 'react';
import { ImageBackground } from 'react-native';
import { headerHeight } from '@themes/ThemeComponent/Common/CommonProps';
import { Metrics } from '@themes';
import { Color } from '@themes/Theme';
import Header, { HeaderProps } from '../../Common/Header';

export interface PrimaryHeaderProps extends HeaderProps{

}
class PrimaryHeader extends PureComponent<PrimaryHeaderProps> {
  render() {
    const { ...otherProps } = this.props;
    return (
      <ImageBackground
        source={require('@assets/images/headerBackground.png')}
        style={{
        }}
        imageStyle={{
          width: Metrics.screenWidth + 2,
          marginLeft: -1,
          resizeMode: 'contain',
          marginTop: -30,
          height: headerHeight + 40,
        }}
      >
        <Header
          {...otherProps}
          backgroundColor="transparent"
          color={Color.white}
        />
      </ImageBackground>
    );
  }
}

export default PrimaryHeader;
