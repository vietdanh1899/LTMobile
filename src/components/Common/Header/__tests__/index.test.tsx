import React from 'react';
import {
  basicTest, testStyleProps, EnumComponent, getStyle, getComponent,
} from '@utils/testHelper';
import { lightTheme } from '@themes';
import { Color } from '@themes/Theme';
import Header from '..';

const target = <Header />;
const type = EnumComponent.HOC;

describe('Header Component', () => {
  basicTest(target, type);

  it('should apply all style-related props', () => {
    const props = {
      position: 'absolute',
      height: 10,
      width: 10,
      top: 10,
      left: 10,
      right: 10,
      bottom: 10,
      borderBottomColor: Color.orange,
      borderBottomWidth: 1,
      backgroundColor: Color.orange,
    };
    testStyleProps(target, props, type, 'containerStyle');
  });

  it('should apply all custom props', () => {
    /**
     * transparent
     */
    const testStyle = getStyle(target, { transparent: true }, type, 'containerStyle');
    expect(testStyle.backgroundColor).toBe('transparent');

    /**
     * backIcon, closeIcon
     */
    const props = getComponent(target, { backIcon: true }, type).props();
    expect(props.leftComponent).toMatchObject({
      icon: 'arrowleft',
      type: 'antdesign',
      size: 25,
      color: lightTheme.Header.leftColor,
    });

    const props2 = getComponent(target, { closeIcon: true }, type).props();
    expect(props2.leftComponent).toMatchObject({
      icon: 'close',
      type: 'antdesign',
      size: 25,
      color: lightTheme.Header.leftColor,
    });

    /**
     * title, logo
     */
    const props3 = getComponent(target, { title: 'Header' }, type).props();
    expect(props3.centerComponent).toMatchObject({
      text: 'Header',
    });

    const props4 = getComponent(target, { logo: true }, type).props();
    expect(props4.centerComponent.props.testID).toBe('centerComponentLogo');

    /**
     * switchTheme
     */
    const props5 = getComponent(target, { switchTheme: true }, type).props();
    expect(props5.rightComponent.props.testID).toBe('rightComponentSwitchTheme');

    /**
     * shadow
     */
    const testStyle2 = getStyle(target, { shadow: true }, type, 'containerStyle');
    expect(testStyle2).toMatchObject(lightTheme.shadowView);

    /**
     * leftColor, centerColor, rightColor, color
     */
    const props6 = getComponent(target, {
      leftColor: Color.orange, backIcon: true,
    }, type).props();
    expect(props6.leftComponent.color).toBe(Color.orange);

    const props7 = getComponent(target, {
      centerColor: Color.orange, title: 'Header',
    }, type).props();
    expect(props7.centerComponent.style.color).toBe(Color.orange);

    const props8 = getComponent(target, {
      rightColor: Color.orange, switchTheme: true,
    }, type).props();
    expect(props8.rightComponent.props.children[0].props.color).toBe(Color.orange);

    const props9 = getComponent(target, {
      color: Color.orange, backIcon: true, switchTheme: true, title: 'Header',
    }, type).props();
    expect(props9.leftComponent.color).toBe(Color.orange);
    expect(props9.centerComponent.style.color).toBe(Color.orange);
    expect(props9.rightComponent.props.children[0].props.color).toBe(Color.orange);
  });
});
