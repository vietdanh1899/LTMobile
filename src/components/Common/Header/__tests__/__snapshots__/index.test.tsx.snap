// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component should render without issues 1`] = `
<ForwardRef(Themed.Header)
  ViewComponent={[Function]}
  containerStyle={
    Object {
      "backgroundColor": "#5760EB",
      "borderBottomColor": undefined,
      "borderBottomWidth": undefined,
      "bottom": undefined,
      "height": 70,
      "left": undefined,
      "paddingBottom": 10,
      "paddingHorizontal": 18,
      "position": undefined,
      "right": undefined,
      "top": undefined,
      "width": "100%",
    }
  }
  leftContainerStyle={
    Object {
      "paddingRight": 0,
    }
  }
  linearGradientProps={
    Object {
      "backgroundColor": "blue",
    }
  }
  placement="center"
/>
`;
